app/services/usps_in_person_proofing/proofer.rb

```diff
@@ -101,18 +101,26 @@ def request_enrollment_code(unique_id)
       end.body
     end
 
-    # Makes a request to retrieve a new OAuth token
-    # and modifies self to store the token and when
-    # it expires (15 minutes).
+    # Makes a request to retrieve a new OAuth token, caches it, and returns it. Tokens have
+    # historically had 15 minute expirys
     # @return [String] the token
     def retrieve_token!
       body = request_token
-      @@token_expires_at = Time.zone.now + body['expires_in']
-      @@token = "#{body['token_type']} #{body['access_token']}"
+      expires_at = Time.zone.now + body['expires_in']
+      token = "#{body['token_type']} #{body['access_token']}"
+      Rails.cache.write(AUTH_TOKEN_CACHE_KEY, token, expires_at: expires_at)
```

Comment:
Should we write our own Redis caching to avoid the odd behavior of the Rails cache stuff?

We could do a `setex` instead of I assume a `set` and then `expireat`.

Reply:
This works in practice and also works in all environments with or without redis which I like. I'm not really concerned about the two separate `set` and `expireat` calls since this only happens every 15 minutes. I think the value of writing something ourselves would be if it's easier to read and maintain, and less likely to break in the future. It's not clear to me that any of those would be the case with writing our own caching logic. What do you think? fwiw I don't actually know what writing our own redis caching would look like here but I'm expecting something like

```
if redis_is_configured # deployed environments
    cache_to_redis(token, expires_at)
else # local dev/tests
    Rails.cache.write(...)
end
```

and then a similar if/else for reading the token.

I do wonder if we could configure our redis server in such a way that it actually respects the `expires_at` and we could get rid of the [ugly workaround.](https://github.com/18F/identity-idp/pull/7580/files#diff-21dd6f37ce8c3f0ded422c037611d8775e409008e522790a272e2b3b4b426091R112-R117) My understanding is that it doesn't work because of how redis is configured but I'm not sure how it needs to be re-configured to actually work

Reply:
If we're calling through to Redis then I'd prefer we use the full capability of [SET](https://redis.io/commands/set/) and include the expiry in the original command so we don't have to worry about cases where the execution halts before setting the expiry.

Reply:
@mitchellhenke we did some experimentation a couple months ago that led to this implementation -- see [this slack](https://gsa-tts.slack.com/archives/C03FA4VBN76/p1667592731100459?thread_ts=1667514804.376859&cid=C03FA4VBN76) post with the TLDR. I'd love to drop the extra code but it seems to be necessary if we use Rails.cache with redis as a backend in its current configuration

Happy to hop on a call and talk through this if you still have concerns/questions. I'm not a fan of this workaround but do want to use redis for caching these auth tokens

Reply:
Ahhh, thank you, yep, makes sense!